hail/src/main/scala/is/hail/utils/package.scala

```diff
@@ -939,6 +941,76 @@ package object utils extends Logging
     }
     offset
   }
+
+  /**
+   * Merge the sorted `IndexedSeq`s `xs` and `ys` using comparison function `lt`.
+   */
+  def merge[A](xs: IndexedSeq[A], ys: IndexedSeq[A], lt: (A, A) => Boolean): IndexedSeq[A] =
+    (xs.length, ys.length) match {
+      case (0, _) => ys
+      case (_, 0) => xs
+      case (n, m) =>
+
+        val res = new ArrayBuffer[A](n + m)
+
+        var i = 0
+        var j = 0
+        while (i < n && j < m) {
+          if (lt(xs(i), ys(j))) {
+            res += xs(i)
+            i += 1
+          } else {
+            res += ys(j)
+            j += 1
+          }
+        }
+
+        for (k <- i until n) {
+          res += xs(k)
+        }
+
+        for (k <- j until m) {
+          res += ys(k)
+        }
+
+        res
+    }
+
+
+  /**
+   * Run (task, key) pairs on the `executor`, returning some `F` of the
+   * failures and an `IndexedSeq` of the successes with their corresponding
+   * key. No guarantees are made for the order of the successes.
+   */
+  def runAll[F[_], A](executor: Executor)
+                     (accum: (F[Throwable], (Throwable, Int)) => F[Throwable])
+                     (init: F[Throwable])
+                     (tasks: IndexedSeq[(() => A, Int)])
+  : (F[Throwable], IndexedSeq[(A, Int)]) = {
+    val completionService = new ExecutorCompletionService[(Try[A], Int)](executor)
+
+    for ((task, k) <- tasks) {
+      completionService.submit(() => (Try(task()), k))
+    }
+
+    var err = init
+    val buffer = new mutable.ArrayBuffer[(A, Int)](tasks.length)
+
+    for (_ <- tasks.indices) {
+      completionService.take.get match {
+        case (Success(v), k) =>
+          buffer += ((v, k))
+
+        case (Failure(t), k) =>
+          err = accum(err, (t, k))
+      }
+    }
```

Comment:
I think this adds results to the buffer as they finish, is that right? We discussed that there doesn't seem to be any benefit over just blocking until all are finished.